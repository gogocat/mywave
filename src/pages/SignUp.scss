.sign-up {
    &__success-message {
        margin-top: 2rem;
        text-align: center;
    }
}