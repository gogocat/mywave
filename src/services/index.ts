import { ISignUpFormValues } from 'interfaces';

function signUpService(data: ISignUpFormValues) {
    // eslint-disable-next-line no-console
    console.log('submit form: ', data);
    return data;
}

export default { signUpService };
