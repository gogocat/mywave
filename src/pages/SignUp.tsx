import { useState } from 'react';
import Avatar from '@mui/material/Avatar';
import Paper from '@mui/material/Paper';
import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';
import LockIcon from '@mui/icons-material/Lock';
import Typography from '@mui/material/Typography';
import SignUpForm from 'components/Forms/SignUpForm';
import Tick from 'components/Tick';

import './SignUp.scss';

function SuccessMessage() {
    return (
        <div className="sign-up__success-message">
            <Tick size="lg" />
            <h3 className="text-center">Success!</h3>
            <p>Thank you registered.</p>
        </div>
    );
}

function SignUp() {
    const [isCompleted, setIsCompleted] = useState(false);

    return (
        <Grid 
            container
            className="sign-up"
            component="main"
            sx={{ height: '100vh' }}>
            <Grid
                item
                xs={false}
                sm={4}
                md={7}
                sx={{
                    backgroundImage: 'url(https://source.unsplash.com/random?wallpapers)',
                    backgroundRepeat: 'no-repeat',
                    backgroundColor: (t) =>
                        t.palette.mode === 'light' ? t.palette.grey[50] : t.palette.grey[900],
                    backgroundSize: 'cover',
                    backgroundPosition: 'center',
                }}
            />
            <Grid item
                xs={12}
                sm={8}
                md={5}
                component={Paper}
                elevation={6}
                square
            >
                <Box
                    sx={{
                        my: 8,
                        mx: 4,
                        display: 'flex',
                        flexDirection: 'column',
                        alignItems: 'center',
                    }}
                >
                    <Avatar sx={{ m: 1, bgcolor: 'secondary.main' }}>
                        <LockIcon />
                    </Avatar>
                    <Typography component="h1"
                        variant="h5">
                        Sign Up
                    </Typography>
                    { isCompleted 
                        ? <SuccessMessage />
                        : <SignUpForm onCompleted={setIsCompleted}/>
                    }
                </Box>
            </Grid>
        </Grid>
    );
}

export default SignUp;
