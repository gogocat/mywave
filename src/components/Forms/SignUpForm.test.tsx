import {
    render, 
    screen 
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import services from 'services';
import SignUp from './SignUpForm';


describe('Sing Up screen', () => {
    it('should renders Sing Up form', () => {
        const {
            asFragment
        } = render(<SignUp />);

        const userNameField = screen.getByLabelText('Email Address');
        const passwordField = screen.getByLabelText('Password');
        const confirmPasswordField = screen.getByLabelText('Confirm password');
        const submitButton = screen.getByText('Sign Up', { selector: 'button' });

        expect(userNameField).toBeInTheDocument();
        expect(passwordField).toBeInTheDocument();
        expect(confirmPasswordField).toBeInTheDocument();
        expect(submitButton).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    it('should validate email field', async()=> {
        const user = userEvent.setup();

        render(<SignUp />);

        const userNameField: HTMLInputElement = screen.getByLabelText('Email Address');

        // assert invaid email address
        await user.type(userNameField, 'abc');
        await userEvent.click(document.body);

        expect(userNameField.value).toBe('abc');
        expect(await screen.findByText('email must be a valid email')).toBeInTheDocument();

        // assert valid email address
        await user.type(userNameField, 'abc@abc.com');
        await userEvent.click(document.body);

        expect(await screen.queryByText('email must be a valid email')).not.toBeInTheDocument();
    });

    it('should validate password field', async()=> {
        const user = userEvent.setup();

        render(<SignUp />);

        const VALID_PASSWORD = '#A3';
        
        const passwordField: HTMLInputElement = screen.getByLabelText('Password');
        const confirmPasswordField: HTMLInputElement = screen.getByLabelText('Confirm password');

        // assert invaid password address
        await user.type(passwordField, '1');
        await userEvent.click(document.body);

        expect(passwordField.value).toBe('1');
        expect(await screen.findByText('Must at least 3 characters')).toBeInTheDocument();

        await user.clear(passwordField);
        await user.type(passwordField, 'abc');
        await userEvent.click(document.body);

        expect(passwordField.value).toBe('abc');
        expect(await screen.findByText('Must contain at least one uppercase')).toBeInTheDocument();

        await user.clear(passwordField);
        await user.type(passwordField, 'Abc');
        await userEvent.click(document.body);
        expect(await screen.findByText('Must contain at least one number')).toBeInTheDocument();

        await user.clear(passwordField);
        await user.type(passwordField, 'A2c');
        await userEvent.click(document.body);
        expect(await screen.findByText('Must contain at least one special character')).toBeInTheDocument();

        await user.clear(passwordField);
        await user.type(passwordField, VALID_PASSWORD);
        await userEvent.click(document.body);

        // assert pasword is valid
        const muiErrorCss = passwordField.closest('div');
        expect(muiErrorCss?.classList?.contains('Mui-error')).toBe(false);

        // assert confirm password field
        await user.type(confirmPasswordField, '123');
        await userEvent.click(document.body);
        expect(await screen.findByText('Passwords must match')).toBeInTheDocument();

        await user.clear(confirmPasswordField);
        await user.type(confirmPasswordField, VALID_PASSWORD);
        await userEvent.click(document.body);
        expect(await screen.queryByText('Passwords must match')).not.toBeInTheDocument();
        expect(confirmPasswordField.value).toBe(passwordField.value);
    });

    it('should submit a vaid form', async()=> {
        const spySignUpService = jest.spyOn(services, 'signUpService');

        const user = userEvent.setup();

        render(<SignUp />);

        const VALID_VALUES = {
            userName: 'test@test.com',
            password: 'A1#',
        };

        const userNameField: HTMLInputElement = screen.getByLabelText('Email Address');
        const passwordField: HTMLInputElement = screen.getByLabelText('Password');
        const confirmPasswordField: HTMLInputElement = screen.getByLabelText('Confirm password');
        const submitButton = screen.getByText('Sign Up', { selector: 'button' });

        await user.type(userNameField, VALID_VALUES.userName);
        await user.type(passwordField, VALID_VALUES.password);
        await user.type(confirmPasswordField, VALID_VALUES.password);
        await userEvent.click(submitButton);

        expect(spySignUpService).toHaveBeenCalledTimes(1);
        expect(spySignUpService).toHaveBeenCalledWith({
            email: VALID_VALUES.userName, 
            password: VALID_VALUES.password,
            confirmPassword: VALID_VALUES.password,
        });

    });
});
