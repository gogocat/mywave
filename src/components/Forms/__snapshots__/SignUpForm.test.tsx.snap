// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sing Up screen should renders Sing Up form 1`] = `
<DocumentFragment>
  <div
    class="sign-up-form"
  >
    <form
      action="#"
    >
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="email"
          id="email-label"
        >
          Email Address
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="email"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="email"
            maxlength="120"
            name="email"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Email Address
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="password"
          id="password-label"
        >
          Password
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="password"
            maxlength="120"
            name="password"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Password
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="confirmPassword"
          id="confirmPassword-label"
        >
          Confirm password
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="confirmPassword"
            maxlength="120"
            name="confirmPassword"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Confirm password
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-1vhaqj4-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Sign Up
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>
</DocumentFragment>
`;
